storage/dataflux/worksteal.go

```diff
@@ -45,16 +56,171 @@ type worker struct {
 	idleChannel   chan int
 	result        *listerResult
 	generation    int64
+	lister        *Lister
 }
 
 // workstealListing is the main entry point of the worksteal algorithm.
 // It performs worksteal to achieve highly dynamic object listing.
-func (c *Lister) workstealListing(ctx context.Context) []*storage.ObjectAttrs {
+// workstealListing creates multiple (parallelism) workers that simultaneosly lists
+// objects from the buckets.
+func (c *Lister) workstealListing(ctx context.Context) ([]*storage.ObjectAttrs, error) {
+	var workerErr []error
+	// Idle channel is used to track number of idle workers.
+	idleChannel := make(chan int, c.parallelism)
+	// Result channel is used to track the results from each worker.
+	result := &listerResult{
+		objects: []*storage.ObjectAttrs{},
+	}
+
+	rs, err := newRangeSplitter(defaultAlphabet)
+	if err != nil {
+		return nil, fmt.Errorf("creating new range splitter: %w", err)
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+	// Initialize all workers as idle.
+	for i := 0; i < c.parallelism; i++ {
+		idleWorker := &worker{
+			goroutineID:   i,
+			startRange:    "",
+			endRange:      "",
+			status:        idle,
+			rangesplitter: rs,
+			result:        result,
+			idleChannel:   idleChannel,
+			generation:    int64(0),
+			lister:        c,
+		}
+		idleChannel <- 1
+		g.Go(func() error {
+			if err := idleWorker.doWorkstealListing(ctx); err != nil {
+				workerErr = append(workerErr, err)
+				return fmt.Errorf("listing worker ID %q: %w", i, err)
+			}
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, fmt.Errorf("failed waiting for multiple workers : %w", err)
+	}
+	if len(workerErr) > 0 {
+		return nil, fmt.Errorf("failure in workers : %v", workerErr)
+	}
+
+	close(idleChannel)
+
+	return result.objects, nil
+}
+
+// doWorkstealListing implements the listing logic for each worker.
+// An active worker lists next page of objects to be listed within the given range and then splits
+// range into two half if there are idle workers. Worker keeps the first of splitted range and passes second half of the work in range channel  for
+// idle workers. It continues to do this until shutdown signal is true.
+// An idle worker waits till it finds work in rangeChannel. Once it finds work, it acts like an active worker.
+func (w *worker) doWorkstealListing(ctx context.Context) error {
+	for {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		// If shutdown signal is true, safely terminate the worker.
+		if w.shutDownSignal() {
+			// If the worker is active, update range channel to store the remaining work.
+			if w.status == active {
+				w.lister.ranges <- &listRange{startRange: w.startRange, endRange: w.endRange}
+				// Worker is now idle.
+				w.status = idle
+			}
+			break
+		}
+
+		// If a worker is idle, sleep for a while before checking the next update.
+		// Worker is active when it finds work in range channel.
+		if w.status == idle {
+			if len(w.lister.ranges) == 0 {
```

Comment:
Why not use a `select`?

```
case <-ctx.Done():
   return ctx.Err()
case newRange := <- w.lister.ranges:
   // do work
default:
   time.Sleep(...)   
```

The shutdown signal could be sent through a channel as well or checked in the default case.

Reply:
For this case, we'll have to get newRange, if the worker is already active, we'll have to put the range back to the channel and then start working. This  might increase wait time for idle workers waiting for work. 

 If we want to use select statement, we can do this I think:

case <-ctx.Done():
   return ctx.Err()
case w.shutdownSignal():
   // shutdown logic
case w.status == idle &&  len(w.lister.ranges) == 0:
   // sleep
default:
   // if idle, pick work. 
   // do work

Reply:
However, this seems ok too. My suggestion if keeping as is would be to change the loop to have a terminating condition: 
`for !w.shutDownSignal()`

And then do everything currently in the `w.shutDownSignal()` conditional at the end of the function.

Reply:
Made shutdown signal as loop terminating condition.